File: preset_dump/src/services/json_generator.py
"""JSON report generator for Minilogue XD program data"""

import json
from typing import Any, Dict
from .program_data import ProgramData


class JsonReportGenerator:
    """Generates JSON reports from Minilogue XD program data"""
    
    @property
    def preferred_file_extension(self) -> str:
        return ".json"
    
    def generate_report(self, program_data: ProgramData) -> str:
        """Generate JSON report from program data"""
        return json.dumps(self._serialize_program_data(program_data), indent=2)
    
    def _serialize_program_data(self, program_data: ProgramData) -> Dict[str, Any]:
        """Convert ProgramData to a JSON-serializable dictionary"""
        result = {}
        
        for field_name, field_value in program_data.__dict__.items():
            if hasattr(field_value, 'value'):
                # Handle enum types
                result[field_name] = field_value.value
            elif hasattr(field_value, 'name'):
                # Handle enum types with name attribute
                result[field_name] = field_value.name
            else:
                # Handle primitive types
                result[field_name] = field_value
        
        return result

File: ts_web/src/json-generator.ts
import { ProgramData } from './program-data';

export class JsonGenerator {
  static generate(programData: ProgramData): string {
    const serialized = this.serializeProgramData(programData);
    return JSON.stringify(serialized, null, 2);
  }

  private static serializeProgramData(programData: ProgramData): Record<string, any> {
    const programResult: Record<string, any> = {};

    for (const [key, value] of Object.entries(programData)) {
      if (typeof value === 'object' && value !== null && 'valueOf' in value) {
        // Handle enum types
        programResult[key] = value.valueOf();
      } else {
        // Handle primitive types
        programResult[key] = value;
      }
    }

    const sequencerResult: Record<string, any> = {};
    const userUnitDescriptions: Record<string, any> = {};
    const userUnitMappings: Record<string, any> = {};

    return {
      "program": programResult,
      "sequencerV2": sequencerResult,
      "userUnitDescriptions": userUnitDescriptions,
      "userUnitMappings": userUnitMappings
    };
  }
}